### ISSUE TITLE
Validation set score depends on rows order (using lambdarank)

### ISSUE THREAD

**Issue by dmirecki** [Microsoft/LightGBM#2091]
## Environment info

Operating System: Ubuntu 18.04

Python version: 3.6

LightGBM version: 2.2.2

## Reproducible examples

```
import numpy as np
import lightgbm as lgb

np.random.seed(10)
X_train = np.random.random((10, 2))
X_test = np.random.random((5, 2))

y_train = np.random.choice([0, 1], 10)
y_test = np.random.choice([0, 1], 5)

data_train = lgb.Dataset(
    X_train,
    label=y_train,
    group=[10]
)
data_test = data_train.create_valid(
    X_test,
    label=y_test,
    group=[5]
)
data_test_reversed = data_train.create_valid(
    X_test[::-1],
    label=y_test[::-1],
    group=[5]
)

param = {
    'num_trees': 1,
    'objective': 'lambdarank',
    'seed': 102836,
    'eval_at': (50, ),
    'max_position': 50,
}

bst = lgb.train(
    param,
    data_train,
    valid_sets=[data_test, data_test_reversed],
)
```
The output is:
```
[1]	valid_0's ndcg@50: 0.543771	valid_1's ndcg@50: 0.919721
```
NDCG on the first and second validation sets are different, while the sets are the same (except the row order). I expected to see the same scores in both sets.

**Comment by StrikerRUS** [comment]
Linking #2036.

@dmirecki Hi!
Can you confirm that the issue is gone if you increase the number of train samples from 10 to 50?
I get the following identical results:
```
[1]	valid_0's ndcg@50: 0.63093	valid_1's ndcg@50: 0.63093
```

I think the reason can be found from (**not**) training logs:

```
[LightGBM] [Warning] There are no meaningful features, as all feature values are constant.
[LightGBM] [Info] Total Bins 0
[LightGBM] [Info] Number of data: 10, number of used features: 0
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
```

Compare with 50 samples:

```
[LightGBM] [Info] Total Bins 36
[LightGBM] [Info] Number of data: 50, number of used features: 2
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
```

ping @guolinke

**Comment by dmirecki** [comment]
Hi @StrikerRUS, thanks for your response.

Unfortunately, if I increase train samples to 50, there is the same problem in the next iterations. Modified example:

```
import numpy as np
import lightgbm as lgb

np.random.seed(10)
X_train = np.random.random((50, 2))
X_test = np.random.random((5, 2))

y_train = np.random.choice([0, 1], 50)
y_test = np.random.choice([0, 1], 5)

data_train = lgb.Dataset(
    X_train,
    label=y_train,
    group=[50]
)
data_test = data_train.create_valid(
    X_test,
    label=y_test,
    group=[5]
)
data_test_reversed = data_train.create_valid(
    X_test[::-1],
    label=y_test[::-1],
    group=[5]
)

param = {
    'num_trees': 3,
    'objective': 'lambdarank',
    'seed': 102836,
    'eval_at': (50, ),
    'max_position': 50,
}

bst = lgb.train(
    param,
    data_train,
    valid_sets=[data_test, data_test_reversed],
)
```

Output:
```
[1]	valid_0's ndcg@50: 0.63093	valid_1's ndcg@50: 0.63093
[2]	valid_0's ndcg@50: 0.63093	valid_1's ndcg@50: 1
[3]	valid_0's ndcg@50: 0.63093	valid_1's ndcg@50: 1
```

Logs:
```
[LightGBM] [Info] Total Bins 36
[LightGBM] [Info] Number of data: 50, number of used features: 2
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
```

**Comment by guolinke** [comment]
@dmirecki 
I think this is an expected behavior in your case.
There are the same scores in your test prediction, but with different label. In this case, the ranking results are not unique, and different can produce different ndcg results.

you can find that by adding following in your code:
```python
p = bst.predict(X_test)
y_p_pair = [(a, b) for (a, b) in zip(p, y_test)]
y_p_reverse = y_p_pair[::-1]
sorted1 = sorted(y_p_pair, key=lambda t: t[0])
sorted2 = sorted(y_p_reverse, key=lambda t: t[0])
```

```
>>> sorted1
[(-0.0414107145059861, 0), (-0.02175834732659826, 0), (-0.008637210242153167, 0), (0.03879043895810194, 0), (0.03879043895810194, 1)]
>>> sorted2
[(-0.0414107145059861, 0), (-0.02175834732659826, 0), (-0.008637210242153167, 0), (0.03879043895810194, 1), (0.03879043895810194, 0)]
```

**Comment by dmirecki** [comment]
You are right, thank you!
